File: backend/app/routers/scripts.py
from fastapi import APIRouter
from sqlmodel import select

from ..static_files import (
    create_script_file,
    delete_script_file,
    get_script_program,
    update_script_program,
)
from ..exceptions.not_found import (
    script_not_found_exception,
    user_not_found_exception,
    user_or_website_not_found_exception,
    website_not_found_exception,
)
from ..database import DatabaseDep
from ..models.database_tables import (
    Annotation,
    Script,
    UnpublishedScript,
    User,
    Website,
)
from ..models.responses import (
    BaseScriptResponse,
    ScriptWithAuthorAndWebsiteResponse,
    ScriptWithProgramResponse,
    SuccessResponse,
)
from ..models.requests import (
    PublishScriptRequest,
    CreateAnnotationRequest,
    UpdateScriptRequest,
)

router = APIRouter(prefix="/scripts", tags=["scripts"])


@router.get("/", response_model=list[ScriptWithAuthorAndWebsiteResponse])
def get_scripts(
    session: DatabaseDep,
) -> list[ScriptWithAuthorAndWebsiteResponse]:
    scripts = session.exec(select(Script)).all()
    return [script.toScriptWithAuthorAndWebsiteResponse() for script in scripts]


@router.post("/", response_model=BaseScriptResponse)
def create_script(
    script: PublishScriptRequest, session: DatabaseDep
) -> BaseScriptResponse:
    author = session.get(User, script.author_id)
    website = session.get(Website, script.website_id)
    if not author or not website:
        raise user_or_website_not_found_exception(
            user_id=script.author_id,
            user_found=bool(author),
            website_id=script.website_id,
            website_found=bool(website),
        )

    filename = create_script_file(program=script.program)
    new_script = Script(
        title=script.title,
        author_id=script.author_id,
        created_at=script.created_at,
        description=script.description,
        website_id=script.website_id,
        script_url=filename,
    )

    session.add(new_script)
    session.commit()
    session.refresh(new_script)

    return new_script.toBaseScriptResponse()


@router.patch("/{script_id}", response_model=SuccessResponse)
def update_script(
    script_id: int, script: UpdateScriptRequest, session: DatabaseDep
) -> SuccessResponse:
    print("Got update script request")
    if not (existing_script := session.get(Script, script_id)):
        raise script_not_found_exception(script_id)

    existing_script.title = script.title
    existing_script.description = script.description
    update_script_program(existing_script.script_url, script.program)

    session.commit()
    session.refresh(existing_script)
    return SuccessResponse(ok=True)


@router.get("/{script_id}", response_model=ScriptWithProgramResponse)
def get_script(script_id: int, session: DatabaseDep) -> ScriptWithProgramResponse:
    if not (script := session.get(Script, script_id)):
        raise script_not_found_exception(script_id)

    program = get_script_program(script.script_url)

    return script.toScriptWithProgramResponse(program)


@router.delete("/{script_id}", response_model=SuccessResponse)
def delete_script(script_id: int, session: DatabaseDep) -> SuccessResponse:
    if not (script := session.get(Script, script_id)):
        raise script_not_found_exception(script_id)

    delete_script_file(script.script_url)

    linked_unpublished_script = session.exec(
        select(UnpublishedScript).where(
            UnpublishedScript.published_script_id == script.id
        )
    ).first()
    if linked_unpublished_script:
        linked_unpublished_script.published_script_id = None

    session.delete(script)
    session.commit()

    return SuccessResponse()


@router.post("/{script_id}/annotate", response_model=SuccessResponse)
def create_annotation(
    script_id: int, annotation: CreateAnnotationRequest, session: DatabaseDep
) -> SuccessResponse:
    if not (script := session.get(Script, script_id)):
        raise script_not_found_exception(script_id)

    new_annotation = Annotation(
        script_id=script.id,
        location=annotation.location,
        description=annotation.description,
    )
    session.add(new_annotation)
    session.commit()

    return SuccessResponse()


@router.get("/user/{user_id}", response_model=list[ScriptWithAuthorAndWebsiteResponse])
def get_user_scripts(
    user_id: int, session: DatabaseDep
) -> list[ScriptWithAuthorAndWebsiteResponse]:
    if not (user := session.get(User, user_id)):
        raise user_not_found_exception(user_id)

    return [script.toScriptWithAuthorAndWebsiteResponse() for script in user.scripts]


@router.get(
    "/website/{website_id}", response_model=list[ScriptWithAuthorAndWebsiteResponse]
)
def get_website_scripts(
    website_id: int, session: DatabaseDep
) -> list[ScriptWithAuthorAndWebsiteResponse]:
    if not (website := session.get(Website, website_id)):
        raise website_not_found_exception(website_id)

    return [script.toScriptWithAuthorAndWebsiteResponse() for script in website.scripts]

File: frontend/src/panel/api_temp/scripts.ts
import axios from 'axios';
import { handleError } from '../models/api_temp/APIError';
import {
  BaseScript,
  PublishScriptRequest,
  Script,
  ScriptWithAuthorAndWebsite,
  UpdateScriptRequest,
} from '../models/api_temp/Script';
import APISuccess from '../models/api_temp/APISuccess';
import APIResponse from '../models/api_temp/APIResponse';

const scriptsEndpoint = axios.create({
  baseURL: 'http://localhost:8000/scripts/',
  timeout: 10000,
  headers: {
    'Content-Type': 'application/json; charset=UTF-8',
  },
});

const getScripts = async (): Promise<
  APIResponse<ScriptWithAuthorAndWebsite[]>
> => {
  try {
    console.log('Making request for scripts');
    const response = await scriptsEndpoint.get('');
    return {
      status: 'Loaded',
      data: response.data as ScriptWithAuthorAndWebsite[],
    };
  } catch (err: unknown) {
    return {
      status: 'Error',
      error: handleError(err),
    };
  }
};

const getScript = async (id: number): Promise<APIResponse<Script>> => {
  console.log('Making request for script');
  try {
    const response = await scriptsEndpoint.get(`${id}`);
    return {
      status: 'Loaded',
      data: response.data as Script,
    };
  } catch (err: unknown) {
    return {
      status: 'Error',
      error: handleError(err),
    };
  }
};

const publishScript = async (
  script: PublishScriptRequest,
): Promise<APIResponse<BaseScript>> => {
  try {
    const response = await scriptsEndpoint.post('', script);
    return {
      status: 'Loaded',
      data: response.data as BaseScript,
    };
  } catch (err: unknown) {
    return {
      status: 'Error',
      error: handleError(err),
    };
  }
};

const updateScript = async (
  script_id: number,
  script: UpdateScriptRequest,
): Promise<APIResponse<APISuccess>> => {
  console.log('Making request for script update');
  try {
    const response = await scriptsEndpoint.patch(`${script_id}`, script);
    return {
      status: 'Loaded',
      data: response.data as APISuccess,
    };
  } catch (err: unknown) {
    return {
      status: 'Error',
      error: handleError(err),
    };
  }
};

const deleteScript = async (id: number): Promise<APIResponse<APISuccess>> => {
  try {
    const response = await scriptsEndpoint.delete(`${id}`);
    return {
      status: 'Loaded',
      data: response.data as APISuccess,
    };
  } catch (err: unknown) {
    return {
      status: 'Error',
      error: handleError(err),
    };
  }
};

const getUserScripts = async (
  userId: number,
): Promise<APIResponse<ScriptWithAuthorAndWebsite[]>> => {
  console.log('Making request for scripts');
  try {
    const response = await scriptsEndpoint.get(`user/${userId}`);
    return {
      status: 'Loaded',
      data: response.data as ScriptWithAuthorAndWebsite[],
    };
  } catch (err: unknown) {
    return {
      status: 'Error',
      error: handleError(err),
    };
  }
};

const getWebsiteScripts = async (
  websiteId: number,
): Promise<APIResponse<ScriptWithAuthorAndWebsite[]>> => {
  console.log('Making request for scripts');
  try {
    const response = await scriptsEndpoint.get(`website/${websiteId}`);
    return {
      status: 'Loaded',
      data: response.data as ScriptWithAuthorAndWebsite[],
    };
  } catch (err: unknown) {
    return {
      status: 'Error',
      error: handleError(err),
    };
  }
};

export {
  getScripts,
  getScript,
  publishScript,
  updateScript,
  deleteScript,
  getUserScripts,
  getWebsiteScripts,
};
